### PR TITLE
[mono] Decompress ICU data during iOS/tvOS app startup

### PR DESCRIPTION
**Pull request by imhameed**
This change adds support for decompressing ICU data files using the lzfse
decompressor built into Apple OSes.

If the data file path passed to `GlobalizationNative_LoadICUData` does not have
`.lzfse` as a suffix, then the contents of this file will be mapped into
shareable memory and passed directly to ICU.

Otherwise, the contents are decompressed using a fixed-size working buffer and
are stored inside a cache file with a name that contains a decimal encoded
representation of the originating compressed file's inode number, modification
time, and size. This filesystem metadata is extremely likely to change if the
contents of the source file ever changes, so there's no need to compute a
checksum of the data to determine if the cache is still valid. If a cache file
with an appropriate name is already present then it is mapped into shareable
memory and passed to ICU. Stale cache files (defined here to be any file with a
filename that ends with "-icudt.dat.uncompressed" that doesn't exactly match
the desired cache file name) are purged on startup.

`icudt.dat` for mobile is 2126 KiB right now; when compressed with lzfse it
shrinks to 675 KiB. On an iPhone SE 1st gen, this takes 4ms to decompress.

The "decompression framework" also supports lz4, zlib, and lzma. They are all
worse either in decompression time (lzma especially + zlib) or compression
ratio (lz4):

| Algorithm | icudt.dat compressed size | decompression time |
| --------- | ------------------------- | ------------------ |
| lz4       | 1031 KiB                  | 2.41 ms            |
| lzfse     | 675 KiB                   | 4.20 ms            |
| zlib      | 659 KiB                   | 9.61 ms            |
| lzma      | 427 KiB                   | 49.20 ms           |

I am not comfortable adding 50ms to app startup time. On this same
iPhone SE 1st gen, the "Contacts" app takes 166.7ms to display app-generated
pixels. This is end-to-end latency, from the first indication that iOS has
recognized my finger tap, to the display displaying any non-placeholder
content. This was measured with with a 240fps camera. A barebones
xamarin-macios app takes 125ms to display a "Hello world!" message.

This uses the filename's suffix to control decompression for simplicity, but it
would also be possible to instead frame compressed data with a very simple
header consisting of a long-enough magic number and a decompressed payload size
that can be used as a buffer sizing hint to the decompression loop.

Miscellany:
- On iOS, the cache-directory appears to be app-specific. Example: `/var/mobile/Containers/Data/Application/0C22E0D1-26CD-46CB-9EBC-6CF55B513ED1/Library/Caches/`.
- https://www.usenix.org/system/files/conference/osdi14/osdi14-paper-pillai.pdf

### CODE REVIEW

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c [hunk 1]
```diff
@@ -98,60 +104,180 @@ static int32_t load_icu_data(void* pData)
...
+    #if defined(USE_APPLE_DECOMPRESSION)
+    {
+        size_t path_len = strlen(path);
+        use_lzfse = ends_with(path, path_len, ".lzfse", sizeof(".lzfse") - 1);
+    }
+    #else
+    #endif
     FILE *fp = fopen(path, "rb");
+    if (fp == NULL) {
+    }
```
Comment (am11) [inline src/native/libs/System.Globalization.Native/pal_icushim_static.c:208]:
```suggestion
    #endif
    FILE *fp = fopen(path, "rb");
    if (fp == NULL)
    {
        goto error;
    }
```

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c [hunk 2]
```diff
@@ -98,60 +104,180 @@ static int32_t load_icu_data(void* pData)
...
         log_shim_error("Unable to read ICU dat file");
-        return ret;
+        goto error;
     }
 
     fclose(fp);
+    fp = NULL;
+
+    #if defined(USE_APPLE_DECOMPRESSION)
+    if (use_lzfse) {
```
Comment (am11) [inline src/native/libs/System.Globalization.Native/pal_icushim_static.c:249]: Maybe just use strcmp and delete use_lzfse and ends_with:
```suggestion
    if (strcmp(strrchr(path, '.'), ".lzfse") == 0)
```

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c [hunk 3]
```diff
@@ -98,60 +104,180 @@ static int32_t load_icu_data(void* pData)
...
+        uncompressed_file_buf = apple_decompress(file_buf, file_buf_size);
+        if (uncompressed_file_buf == NULL) {
+            goto error;
+        }
+        icu_data = uncompressed_file_buf;
+        free(file_buf);
+        file_buf = NULL;
+    }
+    #endif
+    if (icu_data == NULL) {
```
Comment (am11) [inline src/native/libs/System.Globalization.Native/pal_icushim_static.c:259]: nit: `{` on its own line (here and other places).

**Comment by am11** [comment]
CI error was (a warning as error) about `use_lzfse`  being unused variable. We can move strcmp statement down to the only place where it is used, and delete `use_lzfse`.

**Review by rolfbjarne — COMMENTED** [review]
I have a couple of thoughts:

* This allocates a big chunk of memory (the size of the entire ICU, ~2 MB).
* It happens on every app launch.

The first one is a problem in certain cases, because the max memory limit is _very_ low for some app extensions (10mb), and in the past we've used about half of that, leaving the other half for the customers. This will take away 40% of the customer's memory budget in a single allocation.

Suggestion:

Would it be possible to save the decompressed output to a (somewhat temporary) file, then mmap that file in read-only mode? That will make the memory purgeable by the OS, and it won't count towards whatever memory limit the OS enforces.

If you only do this if the file doesn't already exist (matching by timestamp, hashcode, or something else), then you'll also solve the second issue.

**Comment by steveisok** [comment]
@imhameed I think you're going to need to add compression to https://github.com/imhameed/runtime/blob/e516581e282a23fc8340475d8c9860d4c787b5e7/src/tasks/AppleAppBuilder/Templates/CMakeLists.txt.template#L59-L63 in order to get the tests building.

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c [hunk 4]
```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
...
+        size_t bytes_to_write = APPLE_DECOMPRESSION_BUF_SIZE - cs.dst_size;
+        uint8_t *write_cursor = buf;
+        dst_size += bytes_to_write;
+        while (bytes_to_write > 0) {
+            ssize_t bytes_written = write(dst_fd, write_cursor, bytes_to_write);
+            int last_error = errno;
+            if (bytes_written == -1)
+            {
+                if (last_error == EINTR) continue;
+                log_shim_error("apple_decompress_to_fd: Error during write().");
```
Comment (rolfbjarne) [inline src/native/libs/System.Globalization.Native/pal_icushim_static.c:189]: There's no `goto error;` here?

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c [hunk 5]
```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
...
+            }
+            else if (bytes_written == 0)
+            {
+                log_shim_error("apple_decompress_to_fd: write() returned 0.");
```
Comment (rolfbjarne) [inline src/native/libs/System.Globalization.Native/pal_icushim_static.c:193]: Nor here?

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c [hunk 6]
```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
...
+        #if defined(USE_APPLE_DECOMPRESSION)
+        apple_mmap_icu_data(path)
+        #else
+        cstdlib_load_icu_data(path)
+        #endif
+        ;
+
+    if (icu_data == NULL)
+    {
+        log_shim_error("Failed to load ICU data.");
```
Comment (rolfbjarne) [inline src/native/libs/System.Globalization.Native/pal_icushim_static.c:437]: Is this supposed to fall through and call `load_icu_data` even if there's no icu_data?

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c [hunk 7]
```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
...
+            {
+                goto error;
+            }
+            result = renameat(dir_fd, tmp_file, dir_fd, cache_file);
+            fsync(dir_fd);
+            if (result == -1)
+            {
+                log_shim_error("apple_mmap_icu_data: failed to rename %s to %s", tmp_file, cache_file);
+                goto error;
+            }
```
Comment (rolfbjarne) [inline src/native/libs/System.Globalization.Native/pal_icushim_static.c:313]: It doesn't look like you're deleting old files here; maybe iterate over all other "icudt-*.dat.decompressed" files and delete them?

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c (hunk 6 above)
Comment (imhameed) [reply src/native/libs/System.Globalization.Native/pal_icushim_static.c:439]: No. This was an erroneous omission.

File: src/native/libs/System.Globalization.Native/pal_icushim_static.c [hunk 8]
```diff
@@ -98,57 +116,356 @@ static int32_t load_icu_data(void* pData)
...
+        size_t bytes_to_write = APPLE_DECOMPRESSION_BUF_SIZE - cs.dst_size;
+        uint8_t *write_cursor = buf;
+        dst_size += bytes_to_write;
+        while (bytes_to_write > 0) {
+            ssize_t bytes_written = write(dst_fd, write_cursor, bytes_to_write);
+            int last_error = errno;
+            if (bytes_written == -1)
+            {
+                if (last_error == EINTR) continue;
+                log_shim_error("apple_decompress_to_fd: Error during write().");
```
Comment (rolfbjarne) [inline src/native/libs/System.Globalization.Native/pal_icushim_static.c:190]: What will the app do if the write fails (or anything else here fails)? Will it exit/crash, or will it fall back to executing, but without any ICU available (whatever that means)?

One reasonable cause for failure is the device running out of disk space.

Comment (steveisok) [reply src/native/libs/System.Globalization.Native/pal_icushim_static.c:190]: If ICU fails to load, the app will terminate with a call to `Environment::FailFast`

Comment (imhameed) [reply src/native/libs/System.Globalization.Native/pal_icushim_static.c:190]: > What will the app do if the write fails (or anything else here fails)? Will it exit/crash, or will it fall back to executing, but without any ICU available (whatever that means)?
> 
> One reasonable cause for failure is the device running out of disk space.

From testing: the app will continue without ICU data available, so internationalization-related functions will return no data or will fail, exactly as if icudt.dat is not available. We could try decompressing into private memory if writing fails, and only fall back to this degraded failure mode if both disk and memory are unavailable.

**Comment by tarekgh** [comment]
> Hmm - that's odd. I would have expected the app to exit. I don't think this is a mode that we should fall back into.

I agree with @steveisok. The behavior should be deterministic and not surprising to the users. In genral, we always fail fast if anything goes wrong. This gives the user the user to know what is wrong and fix it instead of having the user experience multiple failures during the runtime. We'll need to have a failure message be clear though.

**Comment by SamMonoRT** [comment]
@steveisok @tarekgh  @imhameed  - Given all the testing done on the PR, and the fact this is not going to change any functionality till we make the change in xamain macios to use this, we should create a separate issue for the FastFail case and merge this in the current state.

**Review by steveisok — APPROVED** [review]
I'm fine with a follow up on the behavior.

**Pull request merged**